src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkWasmModule.java

```diff
@@ -0,0 +1,257 @@
+import com.dylibso.chicory.wasm.ChicoryException;
+import com.dylibso.chicory.wasm.types.MemoryLimits;
+import com.dylibso.chicory.wasm.WasmModule;
+import com.google.common.collect.ImmutableList;
+import com.google.devtools.build.docgen.annot.DocCategory;
+import com.google.devtools.build.lib.concurrent.ThreadSafety.Immutable;
+import java.time.Duration;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import net.starlark.java.annot.StarlarkBuiltin;
+import net.starlark.java.annot.StarlarkMethod;
+import net.starlark.java.eval.EvalException;
+import net.starlark.java.eval.Printer;
+import net.starlark.java.eval.Starlark;
+import net.starlark.java.eval.StarlarkValue;
+
+@Immutable
+@StarlarkBuiltin(
+    name = "wasm_module",
+    category = DocCategory.BUILTIN,
+    doc = "A WebAssembly module loaded by <code>repository_ctx.load_wasm()</code>.")
+final class StarlarkWasmModule implements StarlarkValue {
+  private final StarlarkPath path;
+  private final Object origPath;
+  private final WasmModule wasmModule;
+  private final String allocateFn;
+
+  public StarlarkWasmModule(
+      StarlarkPath path,
+      Object origPath,
+      byte[] moduleContent,
+      String allocateFn)
+      throws EvalException {
+    WasmModule wasmModule;
+    try {
+      wasmModule = com.dylibso.chicory.wasm.Parser.parse(moduleContent);
+    } catch (ChicoryException e) {
+      throw new EvalException(e);
+    }
+    validateModule(wasmModule, allocateFn);
+
+    this.path = path;
+    this.origPath = origPath;
+    this.wasmModule = wasmModule;
+    this.allocateFn = allocateFn;
+  }
+
+  @Override
+  public boolean isImmutable() {
+    return true;
+  }
+
+  @Override
+  public void repr(Printer printer) {
+    printer.append("<wasm_module path=");
+    printer.repr(origPath);
+    printer.append(">");
+  }
+
+  public StarlarkPath getPath() {
+    return path;
+  }
+
+  @StarlarkMethod(
+      name = "path",
+      structField = true,
+      doc = "The path this WebAssembly module was loaded from.")
+  public Object getOrigPath() {
+    return origPath;
+  }
+
+  public StarlarkWasmExecutionResult execute(
+      String functionName,
+      byte[] input,
+      Duration timeout,
+      long memLimitBytes)
+      throws EvalException, InterruptedException {
+    var memLimits = getMemLimits(memLimitBytes);
+    // Perform initialization and execution in a separate thread so it can be interrupted
+    // in case of timeout.
+    var wasmThreadFactory =
+        Thread.ofPlatform().name(Thread.currentThread().getName() + "_wasm").factory();
+    StarlarkWasmExecutionResult result;
+    String errMessage;
+    try (var executor = Executors.newSingleThreadExecutor(wasmThreadFactory)) {
+      return executor.invokeAny(
+          ImmutableList.of(() -> run(functionName, input, memLimits)),
```

Comment:
What I'm not entirely sure about is whether this task is interrupted when the current task is interrupted (say via Ctrl+C) the first time. It's definitely interrupted the second time since `ExecutorService#close` calls `shutdownNow` in that case.

Reply:
Interrupt seems to work the first time. I tried executing a function with an infinite loop, saw `bazel(exec-wasm)` using 100% CPU in `htop`, then ctrl-c'd the `bazel` command. It returned to the prompt immediately + that thread went away.

Reply:
Correction, sorry, the `(exec-wasm)` was the label of the entire process because the test repository I'm using is called `exec-wasm`. The process's CPU use goes to zero when interrupted.